src/devices/src/virtio/block/device.rs

```diff
@@ -177,6 +178,41 @@ impl Drop for DiskProperties {
     }
 }
 
+pub(crate) struct BlockIrqTrigger {
+    pub(crate) inner: IrqTrigger,
+    has_batched_vring_irqs: bool,
+}
+
+impl BlockIrqTrigger {
+    fn new() -> std::io::Result<BlockIrqTrigger> {
+        Ok(BlockIrqTrigger {
+            has_batched_vring_irqs: false,
+            inner: IrqTrigger::new()?,
+        })
+    }
+
+    fn do_trigger_vring_irq(&mut self) {
+        self.has_batched_vring_irqs = false;
+        self.inner.trigger_irq(IrqType::Vring).unwrap_or_else(|_| {
+            METRICS.block.event_fails.inc();
+        });
+    }
+
+    fn trigger_vring_irq(&mut self, batch: bool) {
+        if batch {
+            self.has_batched_vring_irqs = true;
+        } else {
+            self.do_trigger_vring_irq();
+        }
+    }
+
+    fn flush_vring_irqs(&mut self) {
+        if self.has_batched_vring_irqs {
+            self.do_trigger_vring_irq();
+        }
+    }
```

Comment:
For better clarity, I would try something like:
```suggestion
    pub fn trigger_vring_irq(&mut self) {
        self.inner.trigger_irq(IrqType::Vring).unwrap_or_else(|_| {
            METRICS.block.event_fails.inc();
        });
    }

    pub fn enqueue_vring_irq(&mut self) {
        self.has_batched_vring_irqs = true;
    }

    pub fn flush_vring_irqs(&mut self) {
        if self.has_batched_vring_irqs {
            self.trigger_vring_irq();
            self.has_batched_vring_irqs = false;
        }
    }
}
```

Then call `enqueue_vring_irq()` or directly `trigger_vring_irq()` in block based on batching flag, instead of passing down batching flag to this object.

Reply:
Removed this code as discussed offline.